src/BoundaryConditions/flat_extrapolation_open_boundary_matching_scheme.jl

```diff
@@ -0,0 +1,108 @@
+using Oceananigans.Grids: xspacings, yspacings, zspacings
+
+"""
+    FlatExtrapolation
+
+Zero gradient perpendicular velocity boundary condition.
+
+For constant spacing:
+```math
+f′(xᵢ) ≈ f′(xᵢ₋₁) + f′′(xᵢ₋₁)(xᵢ₋₁ - xᵢ) + O(Δx²) = f′(xᵢ₋₁) + f′′(xᵢ₋₁)Δx + O(Δx²),
+
+f′(xᵢ₋₁) ≈ (f(xᵢ) - f(xᵢ₋₂)) / 2Δx
+
+f′(xᵢ) ≈ (f(xᵢ) - f(xᵢ₋₂)) / 2Δx + O(Δx) = 0 ∴ f(xᵢ) ≈ f(xᵢ₋₂) + O(Δx)
+
+f′′(xᵢ₋₁) ≈ (f′(xᵢ) - f′(xᵢ₋₂)) / 2Δx = - f′(xᵢ₋₂) / 2Δx ≈ - (f(xᵢ₋₁) - f(xᵢ₋₃)) / (2Δx)²
+
+∴ f(xᵢ) ≈ f(xᵢ₋₂) - (f(xᵢ₋₁) - f(xᵢ₋₃))/2 + O(Δx²)
+```
+
+When the grid spacing is not constant it works out that the factor of 1/2 changes to 
+``Δx₋₁/(Δx₋₂ + Δx₋₃)`` instead.
+"""
+struct FlatExtrapolation{FT}
+    relaxation_timescale :: FT
+end
+
+const FEOBC = BoundaryCondition{<:Open{<:FlatExtrapolation}}
+
+function FlatExtrapolationOpenBoundaryCondition(val = nothing; relaxation_timescale = Inf, kwargs...)
+    classification = Open(FlatExtrapolation(relaxation_timescale))
+    
+    return BoundaryCondition(classification, val; kwargs...)
+end
+
+@inline relax(l, m, c, bc, grid, clock, model_fields) =
+    c + ifelse(isnothing(bc.condition)||!isfinite(clock.last_stage_Δt), 0,
+        (getbc(bc, l, m, grid, clock, model_fields) - c) * min(1, clock.last_stage_Δt / bc.classification.matching_scheme.relaxation_timescale))
+
+@inline spacing_factor(args...) = 1/2
+@inline spacing_factor(Δ::AbstractArray, ::Val{:right}) = @inbounds Δ[end] / (Δ[end-1] + Δ[end-2])
+@inline spacing_factor(Δ::AbstractArray, ::Val{:left})  = @inbounds Δ[1]   / (Δ[2]     + Δ[3]    )
```

Comment:
These wil fail on a non-rectilinear grid

Reply:
Ah I hadn't thought of that, will fix now